### DISCUSSION TITLE
A Very Extensible Template System

### DISCUSSION THREAD

**Discussion by pysan3** [nvim-neorg/neorg#802] [Ideas]
Hi, everybody.

Recently I've been into the `norg` format cuz it's so amazing and easy to write (probly a bit too much tho :P).

As discussed in many places, I believe one feature that is currently lacking with this plugin is an easy way to manage template files.

So I created an extension for norg which not only manages template files but also extends it with the power of [LuaSnip](https://github.com/L3MON4D3/LuaSnip).

### Repo: [Neorg Templates Draft](https://github.com/pysan3/neorg-templates-draft)

Please read the [README.md](https://github.com/pysan3/neorg-templates-draft/blob/main/README.md) (or [`README.norg`](https://github.com/pysan3/neorg-templates-draft/blob/main/README.norg)) for instructions.

## TL; DR

![](https://user-images.githubusercontent.com/41065736/232847417-6f7e32b8-fdd3-4464-9d67-269bb28b363c.gif)

1. `:Neorg templates fload journal` (`force load`: Overwrite buffer without asking)
2. `AUTHOR, TODAY...` will be automatically filled.
3. `{TITLE_INPUT}` is an `input_node` with `{TITLE}` being the default placeholder text.
    - Changes the title.
4. Cursor at `{WEATHER}`.
    - Choosing between options provided from `LuaSnip`'s `choice_node`.
5. Cursor ends up at `{CURSOR}` position.

## Wanting Feedbacks!!

- https://github.com/pysan3/neorg-templates-draft 

As the name of the plugin shows, it is still not ready for use, and on top of that, I'm still wondering whether this concept is a good idea.

I would like to hear any feedback as well as ideas to improve this idea.

Best,
pysan3

**Comment by pysan3** [comment]
Found a bug, updated now... https://github.com/pysan3/neorg-templates-draft/commit/16afdd9f762fc6cef6a6543a821531c6884a3384

**Comment by NTBBloodbath** [comment]
Hey, this looks pretty cool actually. Looking forward to seeing how this module develops, nice job!

Just a small question about this, how good would it play with a custom `:Neorg inject-metadata` implementation? I'm currently working in some changes to core metadata handler in order to make it more extensible (this proposal was already accepted iirc, it was discussed on the Discord server a while ago) so you can have multiple metadata presets like `:Neorg inject-metadata blog-post`. I'll expose presets in the metadata module configuration, in case you find it helpful to reply back.

Cheers!

**Reply by pysan3** [reply]
Thanks for your reply!

At the moment, I haven't got any ideas on how to integrate well with the metadata.

The biggest reason is that this depends on LuaSnip which I would say should not be a dependency for the core plugin and therefore cannot plug this functionality into existing core modules. 

I simply added the `add` sub command where the usecase I was thinking is that you add metadata on your own and then use add command to append the rest of the template. 

Another option would be to define a basic metadata in the template file and use `update-...` afterwards. 

Or another idea. Add a special name keyword `{METADATA:foo}` that automatically calls and is replaced by metagen. 

BTW I'm pretty new to this community so which channel (room?) in discord should I look into to read about that discussion?

**Reply by pysan3** [reply]
@NTBBloodbath 

Added magic keyword `{METADATA}` and `{METADATA:any-subcommand-name}`!

Explanation in [# Magic Keywords](https://github.com/pysan3/neorg-templates-draft#magic-keywords) and implementation [here](https://github.com/pysan3/neorg-templates-draft/blob/a3be9368fe6b462bc7e14c79657f63c2de9d17fb/lua/neorg/modules/external/templates/snippet_handler.lua#L135) if you have interests.
